crypto/threads_pthread.c

```diff
@@ -42,6 +42,404 @@
 #  define USE_RWLOCK
 # endif
 
+static pthread_key_t rcu_thr_key;
+
+struct rcu_thr_data;
+
+/*
+ * users is broken up into 2 parts
+ * bits 0-15 current readers
+ * bit 32-63 - ID
+ */
+# define READER_SHIFT 0
+# define ID_SHIFT 32
+# define READER_SIZE 16
+# define ID_SIZE 32
+
+# define READER_MASK     (((uint64_t)1 << READER_SIZE)-1)
+# define ID_MASK         (((uint64_t)1 << ID_SIZE)-1)
+# define READER_COUNT(x) (((uint64_t)(x) >> READER_SHIFT) & READER_MASK)
+# define ID_VAL(x)       (((uint64_t)(x) >> ID_SHIFT) & ID_MASK)
+# define VAL_READER      ((uint64_t)1 << READER_SHIFT)
+# define VAL_ID(x)       ((uint64_t)x << ID_SHIFT)
+
+/*
+ * This is the core of an rcu lock
+ * it tracks the the readers and writers
+ * for the current quiescence point for
+ * a given lock
+ * users is the 64 bit value that stores
+ * the READERS/ID as defined
+ * above
+ *
+ */
+struct rcu_qp {
+    uint64_t users;
+};
+
+/*
+ * This is the per thread tracking data
+ * that is assigned to each thread participating
+ * in an rcu qp
+ *
+ * qp points to the qp that it last acquired
+ *
+ * count is the threads current recursion count
+ */
+struct rcu_thr_data {
+    struct rcu_qp *qp;
+    int count;
+};
+
+/*
+ * This is the internal version of a CRYPTO_RCU_LOCK
+ * it is cast from CRYPTO_RCU_LOCK
+ */
+struct rcu_lock_internal {
+    CRYPTO_RCU_LOCK lock; /* Must be first */
+    struct rcu_cb_item *cb_items;
+    uint32_t id_ctr;
+    struct rcu_qp *qp_group;
+    size_t group_count;
+    uint32_t reader_idx;
+    uint32_t next_to_retire;
+    uint32_t current_alloc_idx;
+    uint32_t writers_alloced;
+    pthread_mutex_t write_lock;
+    pthread_mutex_t alloc_lock;
+    pthread_cond_t alloc_signal;
+    pthread_mutex_t prior_lock;
+    pthread_cond_t prior_signal;
+};
+
+/*
+ * Called on thread exit to free the pthread key
+ * associated with this thread, if any
+ */
+static void free_rcu_thr_data(void *ptr)
+{
+    struct rcu_thr_data *data = ptr;
+
+# ifdef SANITY_CHECKS
+    if (data->qp != NULL)
+        abort();
+# endif
+    CRYPTO_free(data, __FILE__, __LINE__);
+}
+
+/*
+ * Called from OPENSSL_SSL_init to initalize the
+ * rcu lock system
+ */
+void CRYPTO_THREAD_rcu_init(void)
+{
+    pthread_key_create(&rcu_thr_key, free_rcu_thr_data);
+}
+
+/*
+ * Read side acquisition of the current qp
+ */
+static inline struct rcu_qp *get_hold_current_qp(CRYPTO_RCU_LOCK *lock)
+{
+    uint32_t qp_idx;
+    struct rcu_lock_internal *ilock =
+        (struct rcu_lock_internal *)lock;
+
+    /*
+     * get the current qp index
+     */
+    qp_idx = __atomic_load_n(&ilock->reader_idx, __ATOMIC_SEQ_CST);
+    __atomic_add_fetch(&ilock->qp_group[qp_idx].users, VAL_READER,
```

Comment:
Use of gcc/clang atomic builtins will probably be our biggest obstacle to moving this forward.

Up until now we haven't assumed we have atomics support on a given platform, though I've previously advocated to change this. On threaded builds without atomics we fall back to "emulating" atomics using locks &mdash; see our own existing atomics APIs.

The problem is not actually hardware support &mdash; everything implements higher level synchronisation primitives using atomics, but rather the compiler support. Since there's no standard for atomics (C11 atomics haven't proven popular and we can't rely on C11 policy-wise anyway) it basically entails having #ifdef for every different class of compiler with different atomics builtins. So it means enabling threaded builds for new compilers means adding some code, rather than an automatic fallback to a mutex. Of course in non-threaded builds everything is trivial and just implemented as a normal load/store.

My view is the per-compiler work is manageable and a few compilers will cover the platforms we know about. 

For some background see https://www.openssl.org/blog/blog/2023/07/24/f2f-otc-committers-day3/
in particular the Locking and Atomics sections and links to slides for both there.

Not a concern for Windows interestingly as it provides its own APIs for this.

Reply:
I understand what you are saying here, but it begs the question: How do we ensure coverage for our more esoteric platforms?  Looking at the supported [platforms page](https://www.openssl.org/policies/general-supplemental/platforms.html) I see we are generally covered on almost all of our targets using the gcc/clang builtins or the windows Interlocked API.  Notable exceptions include:
*VSI
*Sun C
*Borland C
*djgpp

djgpp is manageable as its freely available and integrated into our CI.  However Sun/Borland C and VSI have limited availability.  Whats the process for determininig what their atomic api is (more specificially determining if they have a gcc/clang compatibile API or not)?

Reply:
Ok, I've abstracted the atomic operations to a set of macros, and implemented them for gcc/clang. All other platforms will use the (admittedly) horrible fallback implementations, but we can add custom atomic support as platforms demand